File: Wildcard_Matching.cpp
/*44. Wildcard Matching  QuestionEditorial Solution  My Submissions
Total Accepted: 71003 Total Submissions: 383621 Difficulty: Hard
Implement wildcard pattern matching with support for '?' and '*'.

'?' Matches any single character.
'*' Matches any sequence of characters (including the empty sequence).

The matching should cover the entire input string (not partial).

The function prototype should be:
bool isMatch(const char *s, const char *p)

Some examples:
isMatch("aa","a") → false
isMatch("aa","aa") → true
isMatch("aaa","aa") → false
isMatch("aa", "*") → true
isMatch("aa", "a*") → true
isMatch("ab", "?*") → true
isMatch("aab", "c*a*b") → false
Hide Company Tags Google Snapchat Facebook
Hide Tags Dynamic Programming Backtracking Greedy String
Hide Similar Problems (H) Regular Expression Matching
*/



/*linear time    constant space
use two index, i1 for str, i2 for pattern
index for *
match for * at s postion
start checking string
	if p[i2]=='?' or s[i1]==p[i2], increase i1, i2
	else if p[i2]=='*', mark this position as index, match=i1
	else if index!=-1, means exist *, i2=index+1, match++, i1=match
	else return false, character does not match
start checking remaing char in pattern
	while (p < pattern.length() && pattern.charAt(p) == '*') p++;
return p == pattern.length();
	*/

/////////////////////////////////////////////////////////////////////////////////////
//C++
class Solution {
public:
    bool isMatch(string s, string p) {
        int i1 = 0; //point to s
        int i2 = 0; //point to p
        int match = 0;
        int index = -1; 
        while(i1 < s.size()){
            if(i2<p.size() && 
                (p[i2]=='?' || s[i1]==p[i2])){
                    i1++;
                    i2++;
            } else if(i2<p.size() && p[i2]=='*'){
                index = i2;
                match = i1;
                i2++;
            } else if(index!=-1){
                i2 = index+1;
                match++;
                i1 = match;
            } else {
                return false;
            }
        }
        
        while(i2<p.size() && p[i2]=='*'){
            i2++;
        }
        
        return i2==p.size();
        
    }
};



/*DP
O(mn) time and O(mn) space
initialize dp[0][j], j = [1, n]. In this case, s has nothing, to get dp[0][j] = true, p must be '*', '*', '**',etc. Once p.charAt(j-1) != '*', all the dp[0][j] afterwards will be false.

Then start the typical DP loop.
if p[j-1]!='*', dp[i][j] = dp[i-1][j-1] && 
							(s.charAt(i-1)==p.charAt(j-1) || p.charAt(j-1)=='?');
else dp[i][j] = dp[i-1][j] || dp[i][j-1];

Improvement: 
1) optimize 2d dp to 1d dp, this will save space, reduce space complexity to O(N). 
2) use iterative 2-pointer. 2 index*/

/////////////////////////////////////////////////////////////////////////////////////
//Java
public class Solution {
    public boolean isMatch(String s, String p) {
        int m = s.length();
        int n = p.length();
        boolean[][] ret = new boolean[m+1][n+1];
        ret[0][0] = true;
        
        for(int i = 1; i <= m; ++i){
            ret[i][0] = false;
        }
        
        for(int j = 1; j <= n; ++j){
            if(p.charAt(j-1)=='*'){
                ret[0][j] = true;
            } else {
                break;
            }
        }
        
        //start DP
        for(int i = 1; i <= m; ++i){
            for(int j = 1; j <= n; ++j){
                if(p.charAt(j-1)!='*'){
                    ret[i][j] = ret[i-1][j-1] && 
                        (s.charAt(i-1)==p.charAt(j-1) || p.charAt(j-1)=='?');
                } else {
                    ret[i][j] = ret[i-1][j] || ret[i][j-1];
                }
            }
        }
        
        return ret[m][n];
    }
}



public class Solution {
    public boolean isMatch(String s, String p) {
        int i1 = 0; //index for s
        int i2 = 0; //index for p
        int match = 0; //mark if match at s
        int index = -1; //mark * position at p
        while(i1 < s.length()){
            if(i2<p.length() && 
                (p.charAt(i2)=='?' || s.charAt(i1)==p.charAt(i2))){
                    i1++;
                    i2++;
            } else if(i2<p.length() && p.charAt(i2)=='*'){
                index = i2;
                match = i1;
                i2++;
            } else if(index != -1){
                //there is a *
                i2 = index+1;
                match++;
                i1 = match;
            } else {
                return false;
            }
        }
        
        while(i2<p.length() && p.charAt(i2)=='*'){
            i2++;
        }
        
        return i2==p.length();
    }
}

File: Wildcard_Matching.swift
/*
Given an input string (s) and a pattern (p), implement wildcard pattern matching with support for '?' and '*' where:

'?' Matches any single character.
'*' Matches any sequence of characters (including the empty sequence).
The matching should cover the entire input string (not partial).

 

Example 1:

Input: s = "aa", p = "a"
Output: false
Explanation: "a" does not match the entire string "aa".
Example 2:

Input: s = "aa", p = "*"
Output: true
Explanation: '*' matches any sequence.
Example 3:

Input: s = "cb", p = "?a"
Output: false
Explanation: '?' matches 'c', but the second letter is 'a', which does not match 'b'.
Example 4:

Input: s = "adceb", p = "*a*b"
Output: true
Explanation: The first '*' matches the empty sequence, while the second '*' matches the substring "dce".
Example 5:

Input: s = "acdcb", p = "a*c?b"
Output: false
 

Constraints:

0 <= s.length, p.length <= 2000
s contains only lowercase English letters.
p contains only lowercase English letters, '?' or '*'.
*/

/*
Solution 1:
iterate s, p
check if 2 string is match

- According to input, we know that when we see "?" it matches a single character - and when we see "*" it matches a sequence of characters.
- Hence, we could just directly apply the matching approach. We could use two pointers to advance each character in the strings. When the character in p is the same as which in s, or when p is "?", we find a matching and we can advance both pointers. Otherwise, if there is a "*" in p, we can advance p continuously. 
- However, we do have 1 problem here - we need to know where is the last star index. This is critical as if we by any chance have a star pointer in p, then we want to increase the pointer to string S to progressively. This in term means, we need to track, additionally, "the matching character in S where we see p has a *"

Time Complexity: O(max(m, n))
Space Complexity: O(1)
*/
class Solution {
    func isMatch(_ s: String, _ p: String) -> Bool {
        if s.isEmpty, p.isEmpty { return true }
        
        var s = Array(s)
        var p = Array(p)
        
        var indexS = 0
        var indexP = 0
        
        let countS = s.count
        let countP = p.count
        
        var temp = -1
        var match = 0
        while indexS < countS {
            if indexP < countP, 
            (p[indexP] == "?" || p[indexP] == s[indexS]) {
                indexP += 1
                indexS += 1
            } else if indexP < countP, p[indexP] == "*" {
                temp = indexP
                match = indexS
                indexP += 1
            } else if temp != -1 {
                indexP = temp+1
                match += 1
                indexS = match
            } else {
                return false
            }
        }
        
        while indexP < countP, p[indexP] == "*" {
            indexP += 1
        }
        
        return indexP == countP
    }
}